### ISSUE TITLE
Plugin not working? (no abandoned carts)

### ISSUE THREAD

**Issue by Coalescent** [mediabeastnz/craft-commerce-abandoned-cart#32]
Maybe I'm misunderstanding how the plugin works but here's what's happening...

Abandoned Cart 1.3.6
Craft Commerce 2.2.8
Craft 3.3.16.1

I purchased and installed your plugin and put it in test mode. I added some carts with different users and then left them to sit, taking no more actions on those carts. 

After some time had passed (at least an hour for every cart), these carts eventually were listed under Commerce's Inactive Carts page. However, there were no abandoned carts listed on the Abandoned Carts Dashboard. To be clear, I have a few different carts of varying age (just over an hour to a few hours) but none show up - and when clicking the 'Find Abandoned Carts' button I get the message (not surprisingly, I guess) that zero emails were sent.

Further, there are none listed in the `abandonedcart_carts` table either.

I have not set up any crons yet. If I understand correctly I shouldn't need to for the above to work.

Am I missing something? I cannot glean what that might be from your documentation.

As a side note on another issue, I had assumed that your plugin was going to actually install email templates in `abandoned-cart/emails/` but I think it actually doesn't do that and that those template paths in Settings are only for use when you want to point to some other custom templates in another directory. Is that correct?

Thanks.

**Comment by mediabeastnz** [comment]
Hey @Coalescent sounds like you've got everything setup correctly.
Before we debug a bit more are you just using the default settings?

The plugin only triggers if the total is > 0, isn't marked as completed and has an email attached. Other than it should find the cart and then put the email in the queue with the appropriate delay.

I'll setup a local copy using your versions in the next day or two to make sure it's not as bug.

**Comment by Coalescent** [comment]
Thanks @mediabeastnz. So I got up this AM and still no carts so I hit "Find Abandoned Carts" again and it found 2 of the 3 carts (yes, all are > $0, have email, not complete) and it sent first emails off successfully. What is puzzling is that it took so long. The Updated Dates on the two abandoned carts are each exactly 8 hours later than Commerce's updated dates. No idea why... but that would not seem to be a coincidence. I have the first reminder delay set to 1 and second to 12 (and no second emails yet).

As to the third cart still sitting unfound, it has the same email address as one of the found carts. Should that affect the plugin retrieving that other cart?

**Comment by Coalescent** [comment]
I'm running some more tests... as a side note, if I update the column `dateUpdated` in `commerce_orders` (as per your docs) to an earlier hour, those orders disappear from the Commerce admin UI under Active Carts (no idea why - and they do not appear in the abandoned cart UI or table when hitting "Find Abandoned Carts" - or appear in the Commerce Inactive Cart UI).

Oddly, I can still access their detail view in the admin UI - and the correct updated time shows - and I can test with them still. Just curious if this behaviour is the same for you when you do your manual testing as per your docs?

Anyway, running more tests, it does appear that the carts are not found unless I make the `dateUpdated` value more than 8 hours old. anything less than that and no new carts are found.

Is there any chance this 8 hour issue is related to Timezone settings? I happen to be -8 hours UTC.

**Comment by mediabeastnz** [comment]
That will likely be it. I'll play around with some settings and see if I can resolve this.

**Comment by mediabeastnz** [comment]
Hey @Coalescent I've looked into this further and now that I think about it Timezones only really come into play for displaying dates in Craft.

For reference this is the bit of code that does the hard work. 
This query searches for carts that have been inactive for 1hr (same as Craft's inactive cart feature) and no older than 12 hours (to prevent old carts being picked up).

```php
$carts->where('commerce_orders.dateUpdated <= DATE_ADD(NOW(), INTERVAL - '.$start.' HOUR)');

$carts->andWhere('commerce_orders.dateUpdated >= DATE_ADD(NOW(), INTERVAL - '.$end.' HOUR)');

As you can see timezones wont really come into play here unless you've got PHP on a different setting to your database.

**Comment by Coalescent** [comment]
PHP is on the same timezone as Craft, so I guess that's not it. This is all running on my local atm... perhaps I'll update my findings when we push up to staging in the coming weeks. If you have any thoughts in the meantime though, would be keen to hear them.

**Comment by mediabeastnz** [comment]
I'm actually planning on releasing a big update just after xmas which might help with some of these testing / visibility issues so i'll keep you posted. Also if you have any feature requests let me know.

**Comment by Coalescent** [comment]
Ok - good to know. Maybe we'll hold off on testing this too thoroughly until then. The only item that's occured to me is that it would be nice to be able to suppress the second email (maybe that's possible already and I missed it in the docs). Thanks.

**Comment by BrandonJamesBorders** [comment]
@mediabeastnz I've debugged and feel confident this is a timezone issue. This is what I did:

1. Create a new cart, output the id, look up the value in the database. Notice it is UTC. Adjust it back one hour anyway. Verify that the plugin does not recognize my cart as abandoned.
![image](https://user-images.githubusercontent.com/6044898/87704869-05be3100-c76b-11ea-909f-75f0100120a4.png)
2. Investiage the cart service, see you're using MySQL date functions. 
3. Open up MySQL CLI. Run the functions, see the function returns my local time.
![image](https://user-images.githubusercontent.com/6044898/87705070-57ff5200-c76b-11ea-8a7c-2c0b69b5c214.png)
![image](https://user-images.githubusercontent.com/6044898/87705089-5cc40600-c76b-11ea-83d7-f7be46d5d6e3.png)
4. Craft appears to always store as UTC, so I updated the code to account for it. 
```
$UTC = new DateTimeZone("UTC");
$dateUpdatedStart = new DateTime();
$dateUpdatedStart->setTimezone($UTC);
$dateUpdatedStart->sub(new DateInterval('PT'.$start.'H'));

$dateUpdatedEnd = new DateTime();
$dateUpdatedEnd->setTimezone($UTC);
$dateUpdatedEnd->sub(new DateInterval('PT'.$end.'H'));

$carts = Order::find();
$carts->dateUpdated(['and', '<= '.$dateUpdatedStart->format('Y-m-d H:i:s'), '>= '.$dateUpdatedEnd->format('Y-m-d H:i:s')]);
```
5. Plugin not sees my cart. Woot!

Anything I'm missing here or not considering?
